src/main/java/net/wasdev/wlp/common/plugins/util/InstallFeatureUtil.java

```diff
@@ -0,0 +1,559 @@
+        Set<String> result = new HashSet<String>();
+        File serverXml = new File(fileContext, serverFile);
+        List<String> updatedParsedXmls = new ArrayList<String>();
+        updatedParsedXmls.add(serverFile);
+        if (parsedXmls != null){
+            updatedParsedXmls.addAll(parsedXmls);
+        }
+        if (serverXml.exists()) {
+            try {
+                Document doc = new XmlDocument() {
+                    public Document getDocument(File file) throws IOException, ParserConfigurationException, SAXException {
+                        createDocument(file);
+                        return doc;
+                    }
+                }.getDocument(serverXml);
+                Element root = doc.getDocumentElement();
+                NodeList nodes = root.getChildNodes();
+                for (int i = 0; i < nodes.getLength(); i++) {
+                    if (nodes.item(i) instanceof Element) {
+                        Element child = (Element) nodes.item(i);
+                        if ("featureManager".equals(child.getNodeName())) {
+                            result.addAll(parseFeatureManagerNode(child));
+                        }
+                        if ("include".equals(child.getNodeName())){
+                            result = parseIncludeNode(result, child, fileContext, updatedParsedXmls);
+                        }
+                    }
+                }
+            } catch (IOException | ParserConfigurationException | SAXException e) {
+                warn("Failed to parse the xml file " + serverXml + ": " + e.getLocalizedMessage());
+                debug(e);
+            }
+        }
+        return result;
+    }
+    
+    private Set<String> parseIncludeNode(Set<String> origResult, Element node, File fileContext,
+            List<String> updatedParsedXmls) {
+        Set<String> result = origResult;
+        String includeFileName = node.getAttribute("location");
+        if (!updatedParsedXmls.contains(includeFileName)){
+            String onConflict = node.getAttribute("onConflict");
+            boolean onConflictMerge = Pattern.compile(Pattern.quote("MERGE"), Pattern.CASE_INSENSITIVE).matcher(onConflict).find();
+            boolean onConflictReplace = Pattern.compile(Pattern.quote("REPLACE"), Pattern.CASE_INSENSITIVE).matcher(onConflict).find();
+            boolean onConflictIgnore = Pattern.compile(Pattern.quote("IGNORE"), Pattern.CASE_INSENSITIVE).matcher(onConflict).find();
+            if (!(onConflictMerge || onConflictReplace || onConflictIgnore)){
+                onConflictMerge = true;
+            }
+            if (!onConflictIgnore){
+                if (onConflictMerge){
+                    result.addAll(getServerXmlFeatures(fileContext, includeFileName, updatedParsedXmls));
+                } else if (onConflictReplace){
+                    result = getServerXmlFeatures(fileContext, includeFileName, updatedParsedXmls);
```

Comment:
If you are completely replacing the feature list, then the order in which the includes are parsed is very important. I think you're parsing the nodes in the order in which they appear in the document, so the last item will always have priority. It also means that one `replace` can completely overwrite the result of many `merge`s that previously took place.

Please verify that the order in which you are checking the `server.xml`, `overrides` and `defaults` folder, and `include` items is identical to how they are resolved Open Liberty and WLP runtimes.

Reply:
We will provide unit tests to verify the ordering.

Reply:
@ericglau Where is this test? I don't see it yet.